Set environment variables for container/image using function in ldmx-env.sh

bryngemark opened the pull request:
I am updating _ldmx-sw_, here are the details.

### What are the issues that this addresses?
This adds a function to pass environment variables to a docker container or singularity image, using a function `ldmx setenv` which is set up by sourcing `ldmx-env.sh` as usual. This resolves #1089.

## Check List
- [ ] I successfully compiled _ldmx-sw_ with my developments -- N/A, this is not compiled code 

- [x] I ran my developments and the following shows that they are successful.

Environment variables get added to a bash array `LDMX_CONTAINER_ENVS` by using the syntax `ldmx setenv myVariableName=someValue`


``` bash
> source scripts/ldmx-env.sh
> echo $LDMX_CONTAINER_ENVS

> ldmx setenv myVar=53
Added container environment variable myVar=53
> echo $LDMX_CONTAINER_ENVS
myVar=53
```

To test, I made a simple script:
``` bash 
 > echo -e '#!/bin/bash \n\n echo "Printing the environment variable value $myVar" ' > someCode.sh 
```
and running it with the container I get
``` bash 
> ldmx source someCode.sh
Printing the environment variable value 53
```
while if I run outside I get
``` bash
> source someCode.sh
Printing the environment variable value
```

No submodules.

I _have not_ tested yet that this also works with singularity, but I have implemented the needed functionality (I believe). Perhaps @LisaALoman you could try it and let me know?

bryngemark commented:
Oh and besides the singularity check, I will need to rebase this branch.

EinarElen commented:
I don't think the singularity version works just yet. To see this, I tried 
```sh
> ldmx setenv test=bonk # or something more useful
Added container environment variable test=bonk
> echo ${LDMX_CONTAINER_ENVS[@]}
test=bonk
> set -x 
> ldmx ls 
# bunch of output, skipping to relevant parts
+ env_list=',test=bonk '
+ singularity run --no-home --cleanenv --env LDMX_BASE=/scratch/ldmx/einar,DISPLAY=:0 ,test=bonk ' #LDMX_BASE=/scratch/ldmx/einar,DISPLAY=:0' --bind /scratch/ldmx/einar,/tmp/.X11-unix ldmx_dev_latest.sif /scratch/ldmx/einar/ldmx-sw/. ls
FATAL:   could not open image /scratch/ldmx/einar/,test=bonk: failed to retrieve path for /scratch/ldmx/einar/,test=bonk: lstat /scratch/ldmx/einar/,test=bonk: no such file or directory
```

I think this just comes from an extra space + a piece of commented code that is still around :) 
```sh
singularity run --no-home --cleanenv \
      --env LDMX_BASE=${LDMX_BASE},DISPLAY=${LDMX_CONTAINER_DISPLAY}:0 ${env_list} \ #LDMX_BASE=${LDMX_BASE},DISPLAY=${LDMX_CONTAINER_DISPLAY}:0 \ 
```
Changing this to 
``` sh 
    singularity run --no-home --cleanenv \
      --env LDMX_BASE=${LDMX_BASE},DISPLAY=${LDMX_CONTAINER_DISPLAY}:0${env_list} \
      --bind ${csv_list} ${LDMX_SINGULARITY_IMG} "$@"
```
Seems to do the trick. Note that there is no space or comma before `${env_list}` 

Tested with 
``` sh
[einarelen@xi ldmx-sw]$ ldmx env | grep test=     
test=bonk
```
This still doesn't work with multiple environment variables though, I'm not super confident about why but changing from 
```sh 
    for env_to_set in "${LDMX_CONTAINER_ENVS[@]}"; do
```
to 
```sh 
    for env_to_set in ${LDMX_CONTAINER_ENVS[@]}; do
```
does the trick. You can see the problem if you run with -x 
```sh 
+ for env_to_set in "${LDMX_CONTAINER_ENVS[@]}"
+ env_list=',test1=hi test2=bye '
```

LisaALoman commented:
I have tested the singularity version and it works with @EinarElen 's suggestion of changes.

bryngemark commented:
Added a protection agains setting the same variable twice. The following protections have been tested:
1. not using the var=value syntax
``` bash
> ldmx setenv myShitIsBananas
myShitIsBananas doesn't follow the required syntax myEnv=someValue!
Not setting container environment variable myShitIsBananas.
```
2. Trying to set an already correctly set variable name 
``` bash 
> ldmx setenv myShit=bananas
Added container environment variable myShit=bananas
>ldmx setenv myShit=ananas
Already set a variable called myShit;
Try a different name or use 'ldmx clean env' (or re-source the setup script) to clean it.
> ldmx setenv myShot=ananas
Added container environment variable myShot=ananas
```
4. However reusing the same value (previously assigned to another variable) works, as it should.
``` bash 
> ldmx setenv yourShot=bananas
Added container environment variable yourShot=bananas
> ldmx config
[...]
Container Environments: myShit=bananas myShot=ananas yourShot=bananas 
Docker Version: Docker version 20.10.12, build e91ed57
Docker Tag: ldmx/dev:latest
    Digest: ldmx/dev@sha256:e3c0c7c5edc433fe0de306be2576a9579c97568223ceacfdcd0cf0eeb5bcb9e6
```

EinarElen commented:
Regarding 3 above, it might be useful to be able to switch the value of a variable but I'm suspecting that this would be a lot more of a hassle to handle. 

One issue I ran into, if you run `ldmx clean env` it seems to be cleaning up a bit too much 
```
[einarelen@xi ldmx-sw]$ . scripts/ldmx-env.sh 
[einarelen@xi ldmx-sw]$ env | grep LDMX
LDMX_BASE=/scratch/ldmx/einar
LDMX_CONTAINER_DISPLAY=
LDMX_SINGULARITY_IMG=ldmx_dev_latest.sif
[einarelen@xi ldmx-sw]$ ldmx clean env 
[einarelen@xi ldmx-sw]$ env | grep LDMX
LDMX_SINGULARITY_IMG=ldmx_dev_latest.sif
```

bryngemark commented:
> Regarding 3 above, it might be useful to be able to switch the value of a variable but I'm suspecting that this would be a lot more of a hassle to handle.
> 
Yeah I'll pass on that one until someone raises a need

> One issue I ran into, if you run `ldmx clean env` it seems to be cleaning up a bit too much
> 
> ```
> [einarelen@xi ldmx-sw]$ . scripts/ldmx-env.sh 
> [einarelen@xi ldmx-sw]$ env | grep LDMX
> LDMX_BASE=/scratch/ldmx/einar
> LDMX_CONTAINER_DISPLAY=
> LDMX_SINGULARITY_IMG=ldmx_dev_latest.sif
> [einarelen@xi ldmx-sw]$ ldmx clean env 
> [einarelen@xi ldmx-sw]$ env | grep LDMX
> LDMX_SINGULARITY_IMG=ldmx_dev_latest.sif
> ```
> 
Yes that's right, and expected. I can change the suggestion to just re-sourcing the script which cleans env variables and re-sets the display and LDMX_BASE.

> Otherwise, this looks good and is going to be super useful. I'm especially looking forward to setting `ASAN_OPTIONS=detect_leaks=0` when debugging to not get reports about memory leaks 😅
> 
> Edit: Actually, something else seems a bit weird,
> 
> ```
> + local csv_list=/tmp/.X11-unix
> + for dir_to_mount in "${LDMX_CONTAINER_MOUNTS[@]}"
> + csv_list=/scratch/ldmx/einar,/tmp/.X11-unix
> + local env_list=
> + singularity run --no-home --cleanenv --env LDMX_BASE=/scratch/ldmx/einar,DISPLAY=:0 ' '
> FATAL:   could not open image /scratch/ldmx/einar/ : failed to retrieve path for /scratch/ldmx/einar/ : lstat /scratch/ldmx/einar/ : no such file or directory
> + --bind /scratch/ldmx/einar,/tmp/.X11-unix ldmx_dev_latest.sif /scratch/ldmx/einar/ldmx-sw/. ls
> ```
Hmmmm, I'll look into this. Thanks for catching that

bryngemark commented:
Ok @EinarElen could I bother you to test it again? If not, let me know and I'll get to a machine with singularity to test on.

tomeichlersmith commented on scripts/ldmx-env.sh:
```diff
@@ -226,9 +232,13 @@ elif hash singularity &> /dev/null; then
     for dir_to_mount in "${LDMX_CONTAINER_MOUNTS[@]}"; do
       csv_list="$dir_to_mount,$csv_list"
+    done
+	local env_list
```
Just fixing the tabs
```suggestion
    local env_list
```

tomeichlersmith commented on scripts/ldmx-env.sh:
```diff
@@ -360,6 +372,40 @@ __ldmx_base() {
+
+  if [[ $_env_to_set != *"="* ]]; then
+    echo "$_env_to_set doesn't follow the required syntax myEnv=someValue!"
+	echo "Not setting container environment variable $_env_to_set."
```
Just fixing the tabs
```suggestion
	  echo "Not setting container environment variable $_env_to_set."
```

tomeichlersmith reviewed (changes requested):
Need to update to include tab completion:

Line 765: include `setenv` with the other `ldmx` subcommands listed.
Line 770: include `setenv` next to `config` so that it is not completed after it is provided.
Line 798: include `setenv` next to other subcommands so that it is not completed after an env var is provided

bryngemark commented:
> Need to update to include tab completion

Thanks Tom! I hadn't thought of this. It reminds me to let you know that this tab completion is not really working on macOS, because the default version of bash is (very) old:
```
> bash --version
GNU bash, version 3.2.57(1)-release (x86_64-apple-darwin19)
Copyright (C) 2007 Free Software Foundation, Inc.
```
I don't know if this is worth mentioning to users (that updating bash is a great idea). I for one hadn't bothered to deal with it until now when I had to test the functionality after my changes. But now I appreciate having done it.

tomeichlersmith commented:
Thanks for mentioning the old-bash on MacOS. I checked and the [FAQ already talks about this](https://ldmx-software.github.io/docs/faq.html) - did you add it? I can't remember adding this myself.

bryngemark commented:
> Thanks for mentioning the old-bash on MacOS. I checked and the [FAQ already talks about this](https://ldmx-software.github.io/docs/faq.html) - did you add it? I can't remember adding this myself.

Lovely! No I didn't. I suspect PF 😄 
Then I'm all happy with this once it's confirmed that the singularity implementation works.

EinarElen commented:
At 4d7abc0bb534c5616763fedb970114d1f36e255c, still not working right on singularity :( 

```[einarelen@xi ldmx-sw]$ set -x 
++ printf '\033]0;%s@%s:%s\007' einarelen xi /scratch/ldmx/einar/ldmx-sw
[einarelen@xi ldmx-sw]$ ldmx ls
+ ldmx ls
+ [[ 1 == \0 ]]
+ case $1 in
+ __ldmx_run_here ls
+ local _old_pwd=/scratch/ldmx/einar/ldmx-sw/ldmx-g4-comparisons
++ pwd -P
+ local _pwd=/scratch/ldmx/einar/ldmx-sw/.
+ __ldmx_is_mounted /scratch/ldmx/einar/ldmx-sw/.
++ cd /scratch/ldmx/einar/ldmx-sw/.
++ pwd -P
+ local full=/scratch/ldmx/einar/ldmx-sw
+ for _already_mounted in ${LDMX_CONTAINER_MOUNTS[@]}
+ [[ /scratch/ldmx/einar/ldmx-sw/ = /scratch/ldmx/einar/* ]]
+ return 0
+ cd /scratch/ldmx/einar
+ __ldmx_run /scratch/ldmx/einar/ldmx-sw/. ls
+ local csv_list=/tmp/.X11-unix
+ for dir_to_mount in "${LDMX_CONTAINER_MOUNTS[@]}"
+ csv_list=/scratch/ldmx/einar,/tmp/.X11-unix
+ local env_list
+ singularity run --no-home --cleanenv --env LDMX_BASE=/scratch/ldmx/einar,DISPLAY=:0 ' '
FATAL:   could not open image /scratch/ldmx/einar/ : failed to retrieve path for /scratch/ldmx/einar/ : lstat /scratch/ldmx/einar/ : no such file or directory
+ --bind /scratch/ldmx/einar,/tmp/.X11-unix ldmx_dev_latest.sif /scratch/ldmx/einar/ldmx-sw/. ls
bash: --bind: command not found
+ return 127
+ local rc=127
+ cd -
+ export OLDPWD=/scratch/ldmx/einar/ldmx-sw/ldmx-g4-comparisons
+ OLDPWD=/scratch/ldmx/einar/ldmx-sw/ldmx-g4-comparisons
+ return 127
+ return 127
++ printf '\033]0;%s@%s:%s\007' einarelen xi /scratch/ldmx/einar/ldmx-sw
```

bryngemark commented:
@EinarElen it looks like the --bind and image line is not being picked up. I noticed a trailing space on the previous (env setting) line, and just removed it. Want to try this?

EinarElen commented:
Works for one environment variable now, but not multiple 
```
+ singularity run --no-home --cleanenv --env LDMX_BASE=/scratch/ldmx/einar,DISPLAY=:0,foo=bar hi=bye --bind /scratch/ldmx/einar,/tmp/.X11-unix ldmx_dev_latest.sif /scratch/ldmx/einar/ldmx-sw/. ls
FATAL:   could not open image /scratch/ldmx/einar/hi=bye: failed to retrieve path for /scratch/ldmx/einar/hi=bye: lstat /scratch/ldmx/einar/hi=bye: no such file or directory
```

bryngemark commented:
Alright. I am fairly confident this will work now. I have no idea why there were quotation marks in the loop again but now I've removed them.

bryngemark commented:
Yay! Thank you for this somewhat clunky crowd sourced validation :)
